bathbot/src/commands/osu/relax/top.rs

```diff
@@ -0,0 +1,133 @@
+use std::{cmp::Ordering, collections::HashMap};
+
+use bathbot_util::constants::GENERAL_ISSUE;
+use eyre::Result;
+use rosu_v2::{error::OsuError, model::GameMode, request::UserId};
+
+use super::RelaxTop;
+use crate::{
+    active::{
+        ActiveMessages,
+        impls::relax::top::{RelaxTopOrder, RelaxTopPagination},
+    },
+    commands::osu::require_link,
+    core::{Context, commands::CommandOrigin},
+    manager::redis::osu::{UserArgs, UserArgsError},
+};
+
+pub async fn relax_top(orig: CommandOrigin<'_>, args: RelaxTop<'_>) -> Result<()> {
+    top(orig, args).await
+}
+
+pub async fn top(orig: CommandOrigin<'_>, args: RelaxTop<'_>) -> Result<()> {
+    let msg_owner = orig.user_id()?;
+    let mut config = match Context::user_config().with_osu_id(msg_owner).await {
+        Ok(config) => config,
+        Err(err) => {
+            let _ = orig.error(GENERAL_ISSUE).await;
+
+            return Err(err);
+        }
+    };
+
+    let user_id = match user_id!(orig, args) {
+        Some(user_id) => user_id,
+        None => match config.osu.take() {
+            Some(user_id) => UserId::Id(user_id),
+            None => return require_link(&orig).await,
+        },
+    };
+
+    let user_args = UserArgs::rosu_id(&user_id, GameMode::Osu).await;
+
+    let user = match Context::redis().osu_user(user_args).await {
+        Ok(user) => user,
+        Err(UserArgsError::Osu(OsuError::NotFound)) => {
+            let content = match user_id {
+                UserId::Id(user_id) => format!("User with id {user_id} was not found"),
+                UserId::Name(name) => format!("User `{name}` was not found"),
+            };
+
+            return orig.error(content).await;
+        }
+        Err(err) => {
+            let _ = orig.error(GENERAL_ISSUE).await;
+            let err = eyre::Report::new(err).wrap_err("Failed to get user");
+
+            return Err(err);
+        }
+    };
+    let user_id = user.user_id.to_native();
+
+    let client = Context::client();
+    let mut scores = client.get_relax_player_scores(user_id).await?;
+    let player = client.get_relax_player(user_id).await?;
+
+    let map_ids = scores
+        .iter()
+        .take(5)
+        .map(|score| (score.beatmap_id as i32, None))
+        .collect();
+
+    let maps = match Context::osu_map().maps(&map_ids).await {
+        Ok(maps) => maps,
+        Err(err) => {
+            warn!(?err, "Failed to get maps from database");
+
+            HashMap::default()
+        }
+    };
+
+    let player = match player {
+        Some(p) => p,
+        None => {
+            let _ = orig.error("Relax player not found").await;
+
+            return Err(eyre::Report::msg("Relax player not found"));
+        }
+    };
```

Comment:
Good opportunity for the `let Some(player) = .. else {` syntax 😃 
Also I don't think it's necessary to return an error here; a returned `Ok(())` should be fine.